### ISSUE TITLE
Add support for ASP.NET Core's URL fingerprinting syntax

### ISSUE THREAD

**Issue by madskristensen** [webhintio/hint#3794]
# 🚀 Feature request

ASP.NET Core uses a URL parameter to store the file hash, like this:

`/file.js?v=JW_qcvnsPVjH-EA6ctfy9uy8HGXBjPIwokOhCirSWAE`

It would be great to have native support for ASP.NET Core. Right now, ASP.NET Core apps get errors due to having a different fingerprinting/file revving pattern.

I've made a regex that should work. It is `(\?|&)v=[\w-]+`

It will identify the pattern for the following URL strings that I've tested it with:

```
/extensions/fcc-f339fe606-9d51-4fca-895c-d50375137b62/icon.png?v=JW_qcvnsPVjH-EA6ctfy9uy8HGXBjPIwokOhCirSWAE
/test/icon.png?v=JW_qcvnsPVjH-EA6ctfy9uy8HGXBjPIwokOhCirSWAE&hat=4
/test/icon.png?hat=4&v=JW_qcvnsPVjH-EA6ctfy9uy8HGXBjPIwokOhCirSWAE
/test/icon.png?hat=4&v=JW_qcvnsPVjH-EA6ctfy9uy8HGXBjPIwokOhCirSWAE&ost=4
/manifest.json?v=1
```

Let me know if I can help in any way. 

/cc @danroth27 @davidfowl

**Comment by molant** [comment]
Hi @madskristensen,

From the [hint documentation](https://webhint.io/docs/user-guide/hints/hint-http-cache/#what-does-the-hint-check) we link to [this article](https://www.stevesouders.com/blog/2008/08/23/revving-filenames-dont-use-querystring/).

Even though it's from 2008, I think it's still valid. That said, it doesn't take into account that https traffic will (should?) not have this problem. At the same time, I don't think we should be checking for https or not in this rule as that could provide conflicting results when testing locally or in production. And even if this only affects a small percentage of the web, it could still be millions of users...

@Malvoz what do you think? @antross is in paternity leave so don't think he will look at this until his back. Is there any data (http-archive?) that we could look at to help us make a decission?

**Comment by Malvoz** [comment]
The advice in the article is fairly outdated at this point in time. Squid updated their default behavior a little over 10 years ago (source: http://www.squid-cache.org/Versions/v2/2.7/RELEASENOTES.html#s1):

> The default rules to not cache dynamic content from cgi-bin and query URLs have been altered. Previously, the "cache" ACL was used to mark requests as non-cachable - this is enforced even on dynamic content which returns cachability information. This has changed in Squid-2.7 to use the default refresh pattern. Dynamic content is now cached if it is marked as cachable. You should remove the default configuration lines with QUERY (acl, and cache) and replace them with the correct refresh_pattern entries.

If other proxies had the same behavior but didn't change even due to the commotion 10 years ago, then I don't think they're very reliable proxies either way.

**Comment by Malvoz** [comment]
I wasn't able to find anything in the HTTP archive :/

If you look at it like this: Google Fonts is probably one of the most used web services that utilizes query strings in their URLs (`https://fonts.googleapis.com/css?family={..}`), if query string URLs weren't cachable in enough proxies then wouldn't we  have heard from developers for years how fonts weren't being cached?

**Comment by molant** [comment]
I believe Google Fonts don't work correctly in China (or at least I had issues with this in the past).
My main concern is giving an advice that could make the experience worse in countries where the Internet infrastructure might be different to what we are used to.

**Comment by madskristensen** [comment]
@molant I don't believe that had anything to do with URL parameters. Just to be clear, my suggestion here is not that webhint should advice people to put fingerprints in URL parameters, but just recognize that it is a valid and functioning way to fingerprint - such that it won't call it out as missing because it is an unrecognized pattern.

**Comment by hxlnt** [comment]
We chatted about this this morning, and we're thinking we can downgrade this warning to a hint to mitigate the concern while still surfacing the message for those who would like to see it. Thoughts?

**Comment by madskristensen** [comment]
@hxlnt Would that deprive users from getting the satisfaction of getting to green/clean?

**Comment by molant** [comment]
It will depend on what the user has configured as the minimum threshold. IIRC the default one for the browser extension is warning so it will not show up unless they change it.

**Comment by madskristensen** [comment]
@molant in that case it sounds great IMHO.

**Comment by molant** [comment]
This might take a bit because of other priorities but if you want to do a PR we will be happy to help you along the way!

**Comment by madskristensen** [comment]
Sure, can you point me to the right file(s)?

**Comment by molant** [comment]
Sure.
The file to modify would be [this one](https://github.com/webhintio/hint/blob/604b587f6a76d45b1e20200e35b52405fa9817bd/packages/hint-http-cache/src/hint.ts#L430).

The idea will be to detect the finger printing syntax after `cacheRevvingPatterns` fails. If it is detected then change the severity from `Severity.warning` to `Severity.hint` a couple lines later from where I pointed.

Ideally there should be a new test [here](https://github.com/webhintio/hint/blob/main/packages/hint-http-cache/tests/tests.ts) as well 😊

Hopefully the [Contributor Guide](https://webhint.io/docs/contributor-guide/) is clear enough. If it isn't please let us know so we can address that (you can file a new issue or post a comment here and we will open it).

Thank you!

**Comment by molant** [comment]
Hi @madskristensen 

did you have the time to look into this?

**Comment by madskristensen** [comment]
@molant sorry, life happened, and I never got around to doing this. I won't be able to do this in the foreseeable future.

**Comment by madskristensen** [comment]
@molant sorry, life happened, and I never got around to doing this. I won't be able to do this in the foreseeable future.
